addons/account/models/account_account.py

```diff
@@ -811,18 +794,114 @@ def _get_most_frequent_account_for_partner(self, company_id, partner_id, move_ty
+
+        if not name and suggested_accounts:
+            records = self.sudo().browse(suggested_accounts)
+            return _records_to_list(records)
+
+        search_domain = Domain('display_name', 'ilike', name) if name else []
+
+        move_type_accounts = defaultdict(list)
+        move_type_accounts.update({
+            'out': ['income'],
+            'in': ['expense', 'asset'],
+        })
+        move_type_prefix = move_type.split('_')[0]
+        # search all account types if the search contains a number
+        digit_in_search_term = any(c.isdigit() for c in name)
+        internal_group_domain = [('internal_group', 'in', move_type_accounts[move_type_prefix])] if not digit_in_search_term else []
+
+        domain = expression.AND([search_domain, internal_group_domain, domain])
+        records = self.with_context(preferred_account_ids=suggested_accounts).search(domain, limit=limit)
+        return _records_to_list(records, highlight=name)
+
+    @api.model
+    @api.readonly
+    def web_name_search(self, name, specification, domain=None, operator='ilike', limit=100):
+        if self.env.context.get('option_template_ref'):
+            record_tuples = self.name_search(name, domain, operator, limit)
+            return [
+                {
+                    'id': id_,
+                    'display_name': display_name,
+                    'html': html,
+                }
+                for id_, display_name, html in record_tuples
+            ]
+        return super().web_name_search(name, specification, domain, operator, limit)
+
+    def _to_qweb(self, template=None, highlight=None, tag='span', extra_values={}):
+        self.ensure_one()
+        re_search_term = self._get_search_term_regex(highlight)
+        if not template:
+            content = self._highlight(self.display_name, re_search_term) if highlight else self.display_name
+            return Markup(f"<{escape(tag)}>{content}</{escape(tag)}>")
+
+        qweb_content = self.env['ir.qweb']._render(
+            template=template,
+            values={
+                'record': self,
+                **extra_values,
+            },
+        )
+        return self._highlight(qweb_content, re_search_term) if highlight else qweb_content
+
+    @functools.lru_cache
+    def _get_search_term_regex(self, name):
+        return re.compile(fr"{name}", re.IGNORECASE) if name else None
+
+    @api.model
+    def _highlight(self, qweb_content, re_text_to_highlight, highlight_class="text-primary bg-secondary fw-bold"):
+        text_content = str(qweb_content)
+        parts = XML_TAG_REGEX.split(text_content)
+
+        for i, part in enumerate(parts):
+            if not (part.startswith("<") and part.endswith(">")):  # Skip HTML tags
+                parts[i] = re_text_to_highlight.sub(lambda m: f'<span class="{highlight_class}">{escape(m.group(0))}</span>', part)
+
+        return Markup("").join(Markup(part) for part in parts)
```

Comment:
This does not seem like the best idea security wise.

Also, it can highlight the "Suggested" tag

Most, if not all of the changes related to `name_search` should be done client side

Reply:
:facepalm: on the Suggested tag.
I think returning HTML is a good idea for name search, as it can be used throughout odoo without the need to customise every widget. Also @fpodoo wanted the highlighting so maybe we can escape the search text to make it safe?

Reply:
:+1: I'll move the highlighting JS side